hub/util/merge.py

```diff
@@ -0,0 +1,321 @@
+            new_samples_dict,
+            conflict_samples_dict,
+        )
+
+
+def check_common_tensor_mismatches(
+    dataset,
+    target_dataset,
+    tensor_names: Set[str],
+):
+    for tensor_name in tensor_names:
+        target_meta = target_dataset[tensor_name].meta
+        original_meta = dataset[tensor_name].meta
+        original_details = {
+            "htype": original_meta.htype,
+            "sample_compression": original_meta.sample_compression,
+            "chunk_compression": original_meta.chunk_compression,
+        }
+        target_details = {
+            "htype": target_meta.htype,
+            "sample_compression": target_meta.sample_compression,
+            "chunk_compression": target_meta.chunk_compression,
+        }
+        for key, value in original_details.items():
+            if value != target_details[key]:
+                raise MergeMismatchError(tensor_name, key, value, target_details[key])
+
+
+def get_new_indexes(
+    new_elements_ids,
+    target_id_changes_commit_map,
+    target_id_to_index_map,
+):
+    new_indexes = []
+    for id in new_elements_ids:
+        target_id_changes_commit_map.pop(id, None)
+        idx = target_id_to_index_map[id]
+        new_indexes.append(idx)
+    return new_indexes
+
+
+def find_conflicts(
+    original_id_changes_commit_map,
+    target_id_changes_commit_map,
+    original_id_to_index_map,
+    target_id_to_index_map,
+):
+    conflict_indexes = []
+    for id in target_id_changes_commit_map:
+        target_commit_ids = target_id_changes_commit_map[id]
+        original_commit_ids = original_id_changes_commit_map[id]
+        set_original_commit_ids = set(original_commit_ids)
+        idx = None
+        for i, item in enumerate(target_commit_ids):
+            if item in set_original_commit_ids:
+                idx = i
+                break
+
+        # if no id is common or if a commit id other than the most recent commit_id is in common, there's a conflict
+        if idx is None or idx > 0:
+            target_idx = target_id_to_index_map[id]
+            original_idx = original_id_to_index_map[id]
+            conflict_indexes.append((original_idx, target_idx))
+    return conflict_indexes
+
+
+def process_tensor(
+    tensor_name: str,
+    dataset,
+    target_dataset,
+    original_node: CommitNode,
+    target_node: CommitNode,
+    lca_node: CommitNode,
+    new_samples_dict,
+    conflict_samples_dict,
+    conflict_resolution: Optional[str] = None,
+):
+    id_tensor_name = get_sample_id_tensor_key(tensor_name)
+    target_id_tensor = target_dataset[id_tensor_name]
+    original_id_tensor = dataset[id_tensor_name]
+
+    target_id_changes_commit_map = get_changes_commit_ids_for_node(
+        target_dataset, tensor_name, target_node, lca_node
+    )
+
+    original_id_changes_commit_map = get_changes_commit_ids_for_node(
+        dataset, tensor_name, original_node, lca_node
+    )
+
+    original_ids = original_id_tensor.numpy().flatten()
```

Comment:
inherently it could be the case where you have billion rows of dtype=int tensor. Per element we are going to generate uuid4 hash, then call this function. Hence hash itself is going to be much more expensive to store than the integer itself. Though for images this make sense. 

Just wondering have you considered other approaches or what to do per small dtype?

 just throwing e.g. would that make sense to have per row of tensors?

Reply:
essentially your code works for small numbers, but what would happen for large numbers? is there a limitation analysis done (and communicated through docs to users)?

Reply:
Agreed that row-wise would be better for such cases, but that would then limit merge functionality to datasets where all tensor lengths are equal.
As it stands, for large numbers, say a billion rows, this tensor would be 4GB in size, which might be larger than small dtype tensor, such users can opt-out of creating id tensors if the actual data itself has no large images/arrays and give up on merge functionality.
This extra stored data isn't loaded except for any case other than merge.

Reply:
just curious, have we considered to store those additional tensors (not only indices) inside the chunks?

Reply:
I did consider it, but the problem would then be that everytime we're streaming this data say for training, we would end up fetching ids that aren't required.

Reply:
yeah I suspect with tabular data there would be useless overhead, make sense.